flink-end-to-end-tests/test-scripts/test_ha_dataset.sh

```diff
@@ -0,0 +1,139 @@
+
+JM_WATCHDOG_PID=0
+
+# flag indicating if we have already cleared up things after a test
+CLEARED=0
+
+function stop_cluster_and_watchdog() {
+    if [ ${CLEARED} -eq 0 ]; then
+
+        if ! [ ${JM_WATCHDOG_PID} -eq 0 ]; then
+            echo "Killing JM watchdog @ ${JM_WATCHDOG_PID}"
+            kill ${JM_WATCHDOG_PID} 2> /dev/null
+            wait ${JM_WATCHDOG_PID} 2> /dev/null
+        fi
+
+        CLEARED=1
+    fi
+}
+
+function verify_logs() {
+    local OUTPUT=$FLINK_DIR/log/*.out
+    local JM_FAILURES=$1
+    local EXIT_CODE=0
+
+    # verify that we have no alerts
+    if ! [ `cat ${OUTPUT} | wc -l` -eq 0 ]; then
+        echo "FAILURE: Alerts found at the general purpose DataSet job."
+        EXIT_CODE=1
+    fi
+
+    # checks that all apart from the first JM recover the failed jobgraph.
+    if ! [ `grep -r --include '*standalonesession*.log' 'Recovered SubmittedJobGraph' "${FLINK_DIR}/log/" | cut -d ":" -f 1 | uniq | wc -l` -eq ${JM_FAILURES} ]; then
+        echo "FAILURE: A JM did not take over."
+        EXIT_CODE=1
+    fi
+
+    if [[ $EXIT_CODE != 0 ]]; then
+        echo "One or more tests FAILED."
+        exit $EXIT_CODE
+    fi
+}
+
+function jm_watchdog() {
+    local EXPECTED_JMS=$1
+    local IP_PORT=$2
+
+    while true; do
+        local RUNNING_JMS=`jps | grep 'StandaloneSessionClusterEntrypoint' | wc -l`;
+        local MISSING_JMS=$((EXPECTED_JMS-RUNNING_JMS))
+        for (( c=0; c<MISSING_JMS; c++ )); do
+            "$FLINK_DIR"/bin/jobmanager.sh start "localhost" ${IP_PORT}
+        done
+        sleep 1;
+    done
+}
+
+function kill_jm {
+    local JM_PIDS=`jps | grep 'StandaloneSessionClusterEntrypoint' | cut -d " " -f 1`
+    local JM_PIDS=(${JM_PIDS[@]})
+    local PID=${JM_PIDS[0]}
+    kill -9 ${PID}
+
+    echo "Killed JM @ ${PID}"
+}
+
+function run_ha_test() {
+    local PARALLELISM=$1
+
+    local JM_KILLS=3
+
+    CLEARED=0
+    mkdir -p ${TEST_DATA_DIR}/control
+    touch ${TEST_DATA_DIR}/control/test.txt
+
+    # start the cluster on HA mode
+    start_ha_cluster
+
+    echo "Running on HA mode: parallelism=${PARALLELISM}."
+
+    # submit a job in detached mode and let it run
+    local JOB_ID=$($FLINK_DIR/bin/flink run -d -p ${PARALLELISM} \
+        $TEST_PROGRAM_JAR \
+        --loadFactor 4 \
+        --outputPath $TEST_DATA_DIR/out/dataset_allround \
+        --source ${TEST_DATA_DIR}/control/test.txt \
+        | grep "Job has been submitted with JobID" | sed 's/.* //g')
+
+    wait_job_running ${JOB_ID}
+
+    # start the watchdog that keeps the number of JMs stable
+    jm_watchdog 1 "8081" &
+    JM_WATCHDOG_PID=$!
+    echo "Running JM watchdog @ ${JM_WATCHDOG_PID}"
+
+    for (( c=0; c<${JM_KILLS}; c++ )); do
+        # kill the JM and wait for watchdog to
+        # create a new one which will take over
+        kill_jm
+        sleep 20
+        wait_job_running ${JOB_ID}
+    done
+
+    printf "STOP" >> ${TEST_DATA_DIR}/control/test.txt
```

Comment:
Probably we could just for testing the HA. This way though we ensure it completes succesfully after all the restarts.

Reply:
We've already verified that the job is properly restarted in the jobmanager-kill loop, so this should only be about shutting down the job.

Reply:
We've verified that it started being recovered. We check for string "Recovered SubmittedJobGraph" It might happen that it won't be able to run, e.g. it might not be able to allocate resources.

I think with cancel we would not cancel e.g. this one: https://issues.apache.org/jira/browse/FLINK-9635. Am I correct?

Reply:
But the loop executes `wait_job_running ${JOB_ID}`, so don't we know it _actually_ restarted?

Reply:
You are right. Do you think we need the changes in the `Generator` source or shall I maybe just introduce a flag to make the source infinite?